tests/test_ls.py

```diff
@@ -147,12 +145,57 @@ def test_ls_dircache_remove_cached(
     prefix = f"{repository}/{temp_branch}"
     resource = f"{prefix}/"
 
-    try:
-        listing_pre = fs.ls(resource)
-        fs.rm(listing_pre[0]["name"])
+    listing_pre = fs.ls(resource)
+    fs.rm(listing_pre[0]["name"])
 
-        # List again, cache should have been invalidated by rm
-        listing_post = fs.ls(resource)
-        assert len(listing_post) == len(listing_pre) - 1
-    finally:
-        client_helpers.reset_branch(fs.client, repository, temp_branch)
+    # List again, cache should have been invalidated by rm
+    listing_post = fs.ls(resource)
+    assert len(listing_post) == len(listing_pre) - 1
+
+
+def test_ls_dircache_recursive(
+    fs: LakeFSFileSystem,
+    repository: str,
+    temp_branch: str,
+) -> None:
+    prefix = f"{repository}/{temp_branch}"
+
+    # (1) Basic recursive dircache filling
+    fs.dircache.clear()
+    listing = fs.ls(prefix + "/", recursive=True)
+    assert len(fs.dircache) > 1  # Should contain entries for all sub-folders
+
+    # Dircache invariant: all files in an entry must be direct descendants of its parent
+    for cache_dir, files in fs.dircache.items():
+        assert all([fs._parent(v["name"].rstrip("/")) == cache_dir for v in files])
+
+    # (2) Dircache correctness, recursive
+    cached_listing_recursive = fs.ls(prefix + "/", recursive=True)
+    # Recursive listing from cache must contain all items, ordering need not be preserved
+    assert {o["name"] for o in cached_listing_recursive} == {o["name"] for o in listing}
+
+    # (3) Dircache correctness, non-recursive
+    cached_listing_nonrecursive = fs.ls(prefix + "/", recursive=False)
+    # Non-recursive listing from cache most only contain direct descendants of the listed directory
```

Comment:
```suggestion
    # Non-recursive listing from cache must only contain direct descendants of the listed directory
```

Reply:
Addressed in 926fb3be2d1eee23b75dfc9514e8570acff35519